gradle-sls-packaging/src/main/java/com/palantir/gradle/dist/tasks/ResolveProductDependenciesTask.java

```diff
@@ -0,0 +1,347 @@
+import com.palantir.gradle.dist.ProductDependencyReport;
+import com.palantir.gradle.dist.ProductId;
+import com.palantir.gradle.dist.RecommendedProductDependencies;
+import com.palantir.gradle.dist.RecommendedProductDependenciesPlugin;
+import java.io.File;
+import java.io.IOException;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.jar.Manifest;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipFile;
+import org.gradle.api.DefaultTask;
+import org.gradle.api.Project;
+import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.artifacts.component.ComponentIdentifier;
+import org.gradle.api.artifacts.component.ProjectComponentIdentifier;
+import org.gradle.api.artifacts.result.ResolvedComponentResult;
+import org.gradle.api.file.ConfigurableFileCollection;
+import org.gradle.api.file.FileCollection;
+import org.gradle.api.file.RegularFileProperty;
+import org.gradle.api.logging.Logger;
+import org.gradle.api.logging.Logging;
+import org.gradle.api.plugins.JavaPlugin;
+import org.gradle.api.provider.ListProperty;
+import org.gradle.api.provider.Property;
+import org.gradle.api.provider.Provider;
+import org.gradle.api.provider.SetProperty;
+import org.gradle.api.tasks.CacheableTask;
+import org.gradle.api.tasks.Input;
+import org.gradle.api.tasks.OutputFile;
+import org.gradle.api.tasks.TaskAction;
+import org.gradle.jvm.tasks.Jar;
+
+@CacheableTask
+public class ResolveProductDependenciesTask extends DefaultTask {
+    private static final Logger log = Logging.getLogger(ResolveProductDependenciesTask.class);
+
+    private final SetProperty<ProductId> inRepoProductIds =
+            getProject().getObjects().setProperty(ProductId.class);
+    private final Property<String> serviceName = getProject().getObjects().property(String.class);
+    private final Property<String> serviceGroup = getProject().getObjects().property(String.class);
+
+    private final ListProperty<ProductDependency> productDependencies =
+            getProject().getObjects().listProperty(ProductDependency.class);
+    private final SetProperty<ProductId> optionalProductIds =
+            getProject().getObjects().setProperty(ProductId.class);
+    private final SetProperty<ProductId> ignoredProductIds =
+            getProject().getObjects().setProperty(ProductId.class);
+    private final Property<Configuration> productDependenciesConfig =
+            getProject().getObjects().property(Configuration.class);
+
+    private final RegularFileProperty outputFile = getProject().getObjects().fileProperty();
+
+    public ResolveProductDependenciesTask() {
+        dependsOn(otherProjectProductDependenciesTasks());
+        outputFile.convention(() -> new File(getTemporaryDir(), "endpoint-minimum-versions.json"));
```

Comment:
avoid using conventions in favour of property/provider. You should be able to just set the value of the property

Reply:
why? I like conventions in cases where you want to know if set. I don't need it here necessarily, but curious as to why.

Reply:
conventions are generally unsafe and it is considered best practice to use properties instead since they provide a richer API to consume and modify

Reply:
Still confused, sorry. This isn't setting a Convention object (which have been deprecated), but the default value for the property per docs here -https://docs.gradle.org/current/userguide/lazy_configuration.html#applying_conventions.  It's like setting a value, but you can tell if something else changes it.

Reply:
ok, lets keep it as is